cmd/cachecmd/main.go

```diff
@@ -154,24 +155,37 @@ func (c *CacheCmd) fromCacheOrRun(ctx context.Context) (exitcode int, err error)
 		return code, c.updateCacheCmd().Start()
 	}
 
+	var useNativeErr bool
+
 	stdoutf, finallyOut, cancelOut, err := c.prepareCacheFile(stdoutCache)
 	if err != nil {
 		return 0, err
 	}
-	defer func() { err = finallyOut() }()
+	defer func() {
+		errOut := finallyOut()
+		if !useNativeErr {
+			err = errOut
+		}
+	}()
 
 	stderrf, finallyErr, cancelErr, err := c.prepareCacheFile(stderrCache)
 	if err != nil {
 		return 0, err
 	}
-	defer func() { err = finallyErr() }()
+	defer func() {
+		errErr := finallyErr()
+		if !useNativeErr {
```

Comment:
Thank you for the catch!
Maybe it's simpler to ensure that finally() won't return if cancelled.

```diff
diff --git a/cmd/cachecmd/main.go b/cmd/cachecmd/main.go
index ce553c1..8f0eb7a 100644
--- a/cmd/cachecmd/main.go
+++ b/cmd/cachecmd/main.go
@@ -193,17 +193,17 @@ func (c *CacheCmd) prepareCacheFile(path string) (
 	}
 	cancelled := false
 	finally = func() error {
-		// Rename temp file to appropriate file name for cache.
-		if err := tmpf.Close(); err != nil {
-			return fmt.Errorf("failed to close file: %v", err)
-		}
-		// Clean up temp file in case rename failed.
+		// Close and clean up temp file in case rename failed.
+		// Ignore Close() error not to return error when cancelled.
+		tmpf.Close()
 		defer os.Remove(tmpf.Name())
 		if cancelled {
 			// Remove cache file if already exists.
 			os.Remove(path)
 			return nil
 		}
+
+		// Rename temp file to appropriate file name for cache.
 		if err := os.Rename(tmpf.Name(), path); err != nil {
 			return fmt.Errorf("faled to rename: %v", err)
 		}
```